karapace/schema_backup.py

```diff
@@ -44,11 +46,56 @@ class BackupError(Exception):
 
 
 @contextlib.contextmanager
-def Writer(filename: Optional[str] = None) -> Iterable[TextIO]:
-    writer = open(filename, "w", encoding="utf8") if filename else sys.stdout
-    yield writer
-    if filename:
-        writer.close()
+def _writer(file: Union[str, Path], *, overwrite: Optional[bool] = None) -> TextIO:
+    """Opens the given file for writing.
+
+    This function uses a safe temporary file to collect all written data, followed by a final rename. On most systems
+    the final rename is atomic under most conditions, but there are no guarantees. The temporary file is always created
+    next to the given file, to ensure that the temporary file is on the same physical volume as the target file, and
+    avoid issues that might arise when moving data between physical volumes.
+
+    :param file: to open for writing, both the empty string and the conventional single dash ``-`` will yield
+        ``sys.stdout`` instead of actually creating a file for writing.
+    :param overwrite: may be set to ``True`` to overwrite an existing file at the same location.
+    :raises FileExistsError: if ``overwrite`` is not ``True`` and the file already exists.
+    :raises OSError: if writing fails or if the file already exists and is not actually a file.
+    """
+    if file in ("", "-"):
+        yield sys.stdout
+    else:
+        if not isinstance(file, Path):
+            file = Path(file)
+        dst = file.absolute()
+
+        def check_dst() -> None:
+            nonlocal overwrite, dst
+            if dst.exists():
+                if overwrite is not True:
+                    raise FileExistsError(f"--location already exists at {dst}, use --overwrite to replace the file.")
+                if not dst.is_file():
+                    raise FileExistsError(
+                        f"--location already exists at {dst}, but is not a file and thus cannot be overwritten."
+                    )
+
+        check_dst()
+        fd, path = mkstemp(dir=dst.parent, prefix=dst.name)
+        src = Path(path)
+        try:
+            fp = open(fd, "w", encoding="utf8")
+            try:
+                yield fp
+                os.fsync(fd)
```

Comment:
If the file is buffered, flush should be called before fsync. I am not sure if the mkstemp creates buffered file?